README.md
# TS/JS Discogs APIv2 Library Version 0.0.5 

(Starting to add PUT, POST, DELETE endpoints, this will not see a formal release until all are complete)

Tool to talk to Discogs API V2

Currently, Only Supports Personal Access Key.

---
## Installation

This library is libraried on NPM, to install, execute the command :

`npm i js_ts_discogs_api_v2_library`

Add the following to your `.env` file (or create a new one)

Exclude sections in `[]`

`DISCOGS_API_TOKEN=""` [This is Your Personal Access Token, get one here : https://www.discogs.com/settings/developers and click "Generate New Token", Paste the token here]

`DISCOGS_USER_NAME = ""` [Your Discogs UserProfile Name]

`DISCOGS_CURRENCY_MARK = ""` [Options are : `USD`  `GBP` `EUR`  `CAD`  `AUD`  `JPY`  `CHF`  `MXN`  `BRL`  `NZD`  `SEK`  `ZAR` - per Discogs API]

`DISCOGS_PER_PAGE=500` [In Testing 500 is the maximum successful, default is 50, change this to your desired number, 500 or less.]

---
## Usage

After installing via NPM (see top of this file), at the top of your `.ts` file you need to import into your project in the usual way.

We advise to import as `Discogs` but that is of course up to you.

Once imported, you will need to define a variable that you use through your code to call the construct ...   
`const discogs = new Discogs({});`

You are now set up to connect to the Discogs API and start using the data provided.

### To Display RateLimit Information

`console.log(discogs.getRatelimit());`

#### USER INFORMATION

`GET` 
`const user = await discogs.getUser();`

`POST`
`const editUser = await discogs.editUser("[Name]" , "[HomePage]", "[Location]", "[Profile]", "[Currency]");`    
If any of the above need to be changed they should **MUST** be left in order and marked as `""`.  When this happens the field will **NOT** be updated, all other fields will be updated, if a field needs to be blanked, please set to `" "`.

#### SUPPORTED OPTIONS FOR DISCOGS VARIABLES
Some of the variables have a limited set of options defined by Discogs.

`SORT FIELD`
- `label` Sort By Record Label
- `artist` Sort By Artist
- `title` Sort By Release Title
- `catno` Sort By Catalogue Number
- `format` Sort By Format Description
- `rating` Sort By Community Rating
- `added` Sort By Date Added To Collection
- `year` Sort By Release Year

`SORT ORDER`
- `asc` Sort Low to High / A-Z / Old to New
- `desc` Sort High to Low / Z-A / New to Old

#### COLLECTION INFORMATION

- This function will get your collection for further working on, using the values set in your `.env` file, if not included the values are defaulted to : `Page Number`:`1`; `Sort Field`:`added`; `Sort Order`:`desc`.   
`GET` 
`const <Variable> = await discogs.getUserCollection(<Page Number>, <Sort Field>, <Sort Order>);`

- This function will get the `Minimum`, `Median` and `Maximum` Values according to Discogs.    
`GET`
`const <Variable> = await discogs.getUserCollectionValue();`

- This function will get any Custom Fields with their ID number.    
`GET` 
`const <variable> = await discogs.listCustomFields();`

- This function will edit the **CONTENTS** of a custom field and update the field for **EVERY** instance of it, at the current time there is no endpoint to allow creation of Custom Fields or edit Title of Custom Fields.   
`POST`
`const <Variable> = await discogs.editCustomValue(<Release ID>,<Custom Field ID>,<Custom Field Content>);`
- 
#### FOLDER INFORMATION

- This function will get a list of all user created folders    
`GET` 
`const <Variable> = await discogs.getUserFolders();`

- This function will create a new folder.    
`POST` 
`const <Variable> = await discogs.createUserFolder("<New Folder Name>");`

- This function will edit the name of a custom folder. You **CAN NOT** rename either of the default folders: `0`:`Everything` or `1`:`Uncategorised`.    
`POST` 
`const <Variable> = await discogs.editUserFolder("<Folder ID>", "<New Folder Name>");`

- This function will delete a custom folder, if you delete a folder with contents, all the contents will be moved to `1`:`Uncategorised`.    
`DELETE`
`const <Variable> = await discogs.deleteUserFolder("<Folder ID>");`

- This function will get the contents of your collection that are in a specific folder.    
`GET` 
`const <Variable> = await discogs.getUserFolderContents("<Folder ID>");`

- This function adds a new release to your collection and automatically puts it in a folder.     
`POST`
`const <Variable> = await discogs.addReleaseToFolder("<Release ID>, <Folder ID>")`

- This function finds a release in your collection, and will move **EVERY** instance of it into the specified folder.     
`POST`
`const <Variable> = await discogs.changeReleaseFolder("<Release ID>, <Folder ID>")`    
- This function finds a release in your collection, and will move **EVERY** instance of it into the default `Uncategorised` folder (ID `1`).     
`DELETE`
`const <Variable> = await discogs.removeReleaseFromFolder("<Release ID>")`

#### RELEASE INFORMATION

- This function gets the details for the specified Release.   
`GET` 
`const <Variable> = await discogs.getRelease("<Release ID>");`

- This function gets any rating you have set for the specified Release.   
`GET` 
`const <Variable> = await discogs.getReleaseUserRating("<Release ID>}");`

- This function sets your personal rating for the specified Release.    
`POST`
`const <Variable> = await discogs.setChangeReleaseUserRating("<Release ID>","<Rating>");`

- This function deletes your personal rating for the specified Release.   
`DELETE`
`const <Variable> = await discogs.deleteReleaseUserRating("<Release ID>");`

- This function get the global Discogs (Community) rating for the specified Release.    
`GET` 
`const <Variable> = await discogs.getReleaseCommunityRating("<Release ID>");`

- This function gets the `Have` and `Want` Statistics for the specified Release.    
`GET` 
`const <Variable> = await discogs.getReleaseStats("<Release ID>");`

- This function gets the details for the specified Master Release.    
`GET` 
`const <Variable> = await discogs.getMasterRelease("<Master ID>");`

- This function gets the version details for the specified Master Release.    
`GET` 
`const <Variable> = await discogs.getMasterReleaseVersions("<Master ID>");`

#### ARTIST INFORMATION

- This function gets the details for the specified Artist.    
`GET` 
`const <Variable> = await discogs.getArtistDetails("<Artist ID>");`

- This function gets the Releases associated with the specified Artist.    
`GET` 
`const <Variable> = await discogs.getArtistReleases("<Artist ID>");`

#### LABEL INFORMATION

- This function gets the details for the specified Record Label.    
`GET` 
`const <Variable> = await discogs.getLabelDetails("<RecordLabel ID>");`

- This function gets the releases for the specified Record Label.    
`GET` 
`const <Variable> = await discogs.getLabelReleases("<RecordLabel ID>");`

#### WANTLIST INFORMATION

- This function will get the contents of your wishlist / wantlist    
`GET` 
`const <Variable> = await discogs.getUserWantlist();`

- This function will add a release to your wishlist / wantlist.    
`POST`
`const <Variable> = await discogs.addToUserWantlist("<Release ID>");`

- This function will remove a release from your wishlist / wantlist.  It **DOES NOT** add to your collection at the same time, if you want to do this, just chain commands.    
`DELETE`
`const <Variable> = await discogs.removeReleaseFromFolder("<Release ID>");`

# Suggestions, Updates & Errors

If you choose to use this library, and run into a problem, please log an issue or raise a PR, we are committed to this project in the long term and will act on all Issues promptly.

If you have suggestions for further functionality that is not already in the roadmap, please raise an Issue, we will add it to the list.

# Thanks

Thanks go out to [Mike Elsmore](https://github.com/ukmadLz). Without your help and support this project would never have got off the ground.

Thanks go to [Discogs](https://discogs.com) for enabling an API

Thanks go to [Bartve](https://github.com/bartve) for [Disconnect](https://github.com/bartve/disconnect) which although now unsupported, is a great library.  Disconnect was part of the inspiration for this Library.

Thanks go to [Joalla](https://github.com/joalla) for [Discogs_Client](https://github.com/joalla/discogs_client) which was part of the inspiration for this library.  I used this library extensively when I created a Python / Django version of the [VinylVault Website](https://www.thevinylvaultshow.co.uk)

# License 

MIT

# Other Resources

[Discogs API Documentation](https://www.discogs.com/developers/)

[RichCodesWeb on Twitch](https://twitch.tv/RichCodesWeb) - I live stream here most days writing code, it is where the development of this is being completed, as well as other projects.  

Come ask me questions, and / or make suggestions in livechat to shape further development of this library and add-on modules.

CHANGELOG.md
#Changelog

Everything here has been on the `Roadmap` at somepoint, even if it hasn't made it to a published version of that file.  

Please check that file if you feel there is something missing from the project. 

`v0.0.1`
- Getting Started :D
- Added `GET` endpoints for all documented API Endpoints
- Added `README.md`

`v0.0.2`
- Added Ability to Automagically rate limit in 2 ways
- - Detect `429`
- - Check on `Remaining Requests`
- If triggered, RNG number of seconds to wait, (between 10 and 60) and sleep, then continue with latest request.

`v0.0.3`
- 1st NPM release
- Added Search, multiple endpoints plus general search
- `README.md` Updates to reflect changes.

`v0.0.4` (Internal Release Only)
- Preparing for v0.1.0 release
- Starting to Add `PUT` `POST` `DELETE` Endpoints

`v0.0.5` (Internal Release Only)
- Rewrite `README.md` to be more clear
- Seperated RoadMap Section in to `CHANGELOG.md` and `ROADMAP.md`

test/index.js
require("dotenv");
const Discogs = require("../dist").default;
// console.log("Bollocks, This Works, I'm A Software Engineer!");

const main = async () => {
  const discogs = new Discogs({});

    // USER INFORMATION
    // const user = await discogs.getUser();
    // console.log(user.data);

    // const collection = await discogs.getUserCollection("1", "artist", "desc");
    // const collection2 = await discogs.getUserCollection("1", "added", "desc");
    // const wantlist = await discogs.getUserWantlist();
    // const folders = await discogs.getUserFolders();
    // const folderContents = await discogs.getUserFolderContents("3784660", "2");
    const collectionValue = await discogs.getUserCollectionValue();

    // FOLDER INFORMATION


    // COLLECTION INFORMATION


    // RELEASE INFORMATION

    // const releaseDetails = await discogs.getRelease("249504"); //RICK-ROLL   829032
    // const releaseUserRating = await discogs.getReleaseUserRating("829032");

    // const releaseCommunityRating = await discogs.getReleaseCommunityRating("829032");
    // const releaseStats = await discogs.getReleaseStats("249504");
    // const releaseMaster = await discogs.getMasterRelease("96559");
    // const releaseMasterVersions = await discogs.getMasterReleaseVersions("96559");

    // ARTIST INFORMATION

    // const artistDetails = await discogs.getArtistDetails("72872"); //RICK ASTLEY
    // const artistReleases = await discogs.getArtistReleases("72872");

    // LABEL INFORMATION

    // const labelDetails = await discogs.getLabelDetails("157"); //RCA
    // const labelReleases = await discogs.getLabelReleases("895");

    // DISPLAY STUFF ON SCREEN TO TEST

  
    // console.log(folders.data);
    // console.log(collection.data);
    // console.log(wantlist.data);
    // console.log(folderContents.data);
    console.log(collectionValue.data);
    // console.log(releaseDetails.data);
    // console.log(releaseUserRating.data);
    // console.log(releaseUserRating.headers);
    // console.log(releaseCommunityRating.data);
    // console.log(releaseStats.data);
    // console.log(releaseMaster.data);
    // console.log(releaseMasterVersions.data);
    // console.log(artistDetails.data);
    // console.log(artistReleases.data);
    // console.log(labelDetails.data);
    // console.log(labelReleases.data);
    // console.log(discogs.getRatelimit());

    // const createUserFolder = await discogs.createUserFolder("ZzZ Something So Wrong");
    // console.log(createUserFolder);

    // const editFolder = await discogs.editUserFolder("4456646", "ZzZ Something Went Right");
    // console.log(editFolder);

    // const deleteFolder = await discogs.deleteUserFolder("4456646");
    // console.log(deleteFolder);

    // const addReleaseToFolder = await discogs.addReleaseToFolder("23035481", "1");
    // console.log(addReleaseToFolder);

    // const changeReleaseFolder = await discogs.changeReleaseFolder("1773118", "4456646");
    // console.log(changeReleaseFolder);

    // const removeReleaseFromFolder = await discogs.removeReleaseFromFolder("829032");
    // console.log(removeReleaseFromFolder);

    // const editUser = await discogs.editUser("" , "", "", "", "GBP");
    // console.log(editUser);

    // const listCustomFields = await discogs.listCustomFields();
    // console.log(listCustomFields.data);

    // const editCustomFieldValue = await discogs.editCustomFieldValue("829032", "4", "farty noises");
    // console.log(editCustomFieldValue.headers);
    // console.log(editCustomFieldValue.data);

    // const addToWantlist = await discogs.addToUserWantlist("10053468");
    // console.log(addToWantlist.data);
    // console.log(addToWantlist.headers);

    // const addReleaseToCollection = await discogs.removeFromUserCollection("16102732");
    // console.log(addReleaseToCollection.data);
    // console.log(addReleaseToCollection.headers);

    // const setChangeReleaseUserRating = await discogs.setChangeReleaseUserRating("829032", 2);
    // console.log(setChangeReleaseUserRating.data);
    // console.log(setChangeReleaseUserRating.headers);

    // const deleteReleaseUserRating = await discogs.deleteReleaseUserRating("829032");
    // console.log(deleteReleaseUserRating.data);
    // console.log(deleteReleaseUserRating.headers);

};

main();


src/index.ts
import Dotenv from 'dotenv';
import Fetch from 'node-fetch';
import { info, debug, log } from './utils/debug';

Dotenv.config();

info('JS TS Discogs API v2 Library Version 0.0.4a')
info('© Dex Vinyl 2022')
info('Released under MIT License')

interface Ratelimit {
    ratelimit: number,
    remaining: number,
    used: number,
}

export default class Client {
    private protocol = 'https';
    private host: string;
    private port: number;
    private userAgent: string;
    private auth: string|null;
    private ratelimit: Ratelimit;
    private discogsUserName: string;
    private perPage: string;

    private defaults = {
        host: 'api.discogs.com',
        port: 443,
        userAgent: 'JSTSDiscogsAPIV2Library',
        apiVersion: 'v2',
        outputFormat: 'discogs',    // Possible values: 'discogs' / 'plaintext' / 'html'
        requestLimit: 25,           // Maximum number of requests to the Discogs API per interval
        requestLimitAuth: 60,       // Maximum number of requests to the Discogs API per interval when authenticated
        requestLimitInterval: 60000, // Request interval in milliseconds
        discogsUserName:' ', // Default Username can only be set in ENV file
        perPage:'50',
    }

    constructor({
                    host,
                    port,
                    userAgent,
                    token,
                    key,
                    secret,
                    discogsUserName,
                }: {
                    host?: string,
                    port?: number;
                    userAgent?: string;
                    token?: string,
                    key?: string,
                    secret?: string,
                    discogsUserName?: string,
        }
    )
    {
        this.discogsUserName = process.env.DISCOGS_USER_NAME || this.defaults.discogsUserName;
        this.host = host || this.defaults.host;
        this.port = port || this.defaults.port;
        this.userAgent = userAgent || this.defaults.userAgent;
        this.auth = this.createAuthString({ token, key, secret });
        this.ratelimit = {
            ratelimit: 25,
            remaining: 25,
            used: 0,
        };
        this.perPage = process.env.DISCOGS_PER_PAGE || this.defaults.perPage;
    }

    private createAuthString({
                                token,
                                key,
                                secret,
                             }: {
                                token?: string,
                                key?: string,
                                secret?: string,
                                })
    {
        let authString;

        let discogsToken = token || process.env.DISCOGS_API_TOKEN
        let discogsKey = key || process.env.DISCOGS_API_KEY
        let discogsSecret = secret || process.env.DISCOGS_API_SECRET

        if (discogsToken) {
            authString = `token=${(discogsToken)}`;
        } else if ((discogsKey) && (discogsSecret)) {
            authString = `key=${(discogsKey)}, secret=${(discogsSecret)}`;
        }

        return authString || null;
    }

    private async request(path: string, method: string = 'GET', body?: any) {
        const requestHeaders: any = {
            'User-Agent': this.userAgent,
        };

        if (this.auth) {
            requestHeaders['Authorization'] = `Discogs ${this.auth}`
            requestHeaders['Content-Type'] = `application/json`
        }
        while (true) {
            try {
                const fetchObject: any = {
                    method,
                    headers: requestHeaders,
                }
                if(body) {
                    if (typeof body == 'object') {
                        fetchObject.body = JSON.stringify(body);
                    } else {
                        fetchObject.body = body
                    }
                }
                const response = await Fetch(`${this.protocol}://${this.host}/${path}`, fetchObject);
                console.log("RESPONSE CODE: " + response.status);
                const responseCode = String(response.status);
                if (responseCode == "204") {
                    process.exit(0);
                }
                const responseHeaders = response.headers;
                const data = await response.json();
                this.ratelimit = {
                    ratelimit: Number(responseHeaders.get('x-discogs-ratelimit')),
                    remaining: Number(responseHeaders.get('x-discogs-ratelimit-remaining')),
                    used: Number(responseHeaders.get('x-discogs-ratelimit-used'))
                }

                return {
                    data,
                    headers: responseHeaders,
                };
            } catch (error) {
                console.error(error);
                const theError:any = error;
                if (theError.type == "invalid-json") {
                    log("Invalid JSON Received Waiting 5 Seconds Before Retry");
                    await this.delay(5000);
                    debug("Trying Again")
                }
            }
        }
    }

    public getRatelimit(): Ratelimit {
        return this.ratelimit;
    }

    // Helper Functions
    public async getRequest (path: string) {
        return this.request(path, 'GET');
    }
    public async deleteRequest (path: string) {
        return this.request(path, 'DELETE');
    }
    public async postRequest(path: string, data: any) {
        return this.request(path, 'POST', data);
    }
    public async putRequest(path: string, data:any) {
        return this.request(path, 'PUT', data);
    }

//
// HAVE A NAP
//

    private async delay(ms: number) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }

//
// CALCULATE RATE LIMIT
//

    public async calculateRateLimitRemaining(){
        let rateRemaining = this.getRatelimit().remaining;
        let rateBarrier = 2;//this.getRatelimit().ratelimit/5;
        log ("You Have " + rateRemaining + " Requests Remaining");
        if (rateRemaining <= rateBarrier) {
            log("Generating Random Wait Time");
            const rngWaitTime = Math.floor(Math.random() * (55 - 15 + 1)) + 15;
            log("You've Used ALL You API Rate Allowance, Waiting for " + rngWaitTime + " Seconds");
            log ("The Threshold is " + rateBarrier + " Requests");
            log(new Date());
            let rngWaitTimeMS = rngWaitTime * 1000;
            await this.delay(rngWaitTimeMS);
            log ("Ok I Waited " + rngWaitTime + " Seconds, Continuing... ");
            log (new Date());
            await this.delay(1000);
        }
    }

//
// USER SPECIFIC ENDPOINTS
//

    public async getUser() {
        await this.calculateRateLimitRemaining();

        return this.getRequest(`users/${this.discogsUserName}`);
    }

    public async editUser(name: string, homepage: string, location: string, profile: string, currency: string) {
        await this.calculateRateLimitRemaining();

        const userDetails = await this.getRequest(`users/${this.discogsUserName}`);
        console.log(userDetails);

        if (name == "") {
            name = userDetails.data.name;
        }

        if (homepage == "") {
            homepage = userDetails.data.home_page;
        }

        if (location == "") {
            location = userDetails.data.location;
        }

        if (profile == "") {
            profile = userDetails.data.profile;
        }

        if (currency == "") {
            currency = userDetails.data.curr_abbr;
        }

        return this.postRequest(`users/${this.discogsUserName}`, {name:name, home_page:homepage, location:location, profile:profile, curr_abbr:currency});
    }

//
// FOLDER ENDPOINTS
//

    public async getUserFolders() {
        await this.calculateRateLimitRemaining();

        return this.getRequest(`users/${this.discogsUserName}/collection/folders`);
    }

    public async createUserFolder(name: string) {
        await this.calculateRateLimitRemaining();
    
        return this.postRequest(`users/${this.discogsUserName}/collection/folders`, {name: name});
    }

    public async editUserFolder(folder: string, name: string) {
        await this.calculateRateLimitRemaining();

        return this.postRequest(`users/${this.discogsUserName}/collection/folders/${folder}`, { name: name });
    }

    public async deleteUserFolder(folderID: string) {
        await this.calculateRateLimitRemaining();

        return this.deleteRequest(`users/${this.discogsUserName}/collection/folders/${folderID}`);
    }

    public async getUserFolderContents(folder:string, pageNumber:string, sort:string, sortOrder:string) {
        await this.calculateRateLimitRemaining();

        if (!pageNumber){
            pageNumber="1"
        }

        if (!sortOrder){
            sortOrder="desc"
        }

        if (!sort) {
            sort = "added"
        }
        else if (sort == "year") {}
        else if (sort == "artist") {}
        else if (sort == "title") {}
        else if (sort == "catno") {}
        else if (sort == "format") {}
        else if (sort == "rating") {}
        else{
            sort = "added"
        }

        return this.getRequest(`users/${this.discogsUserName}/collection/folders/${folder}/releases?sort=${sort}&sort_order=${sortOrder}&per_page=${this.perPage}&page=${pageNumber}`);
    }

    public async addReleaseToFolder(releaseId:string, folderId:string) {
        await this.calculateRateLimitRemaining();

        return this.postRequest(`users/${this.discogsUserName}/collection/folders/${folderId}/releases/${releaseId}`, {});
    }

    public async changeReleaseFolder(releaseId: string, folderId: string) {
        await this.calculateRateLimitRemaining();

        const getReleaseInfoFromCollection = await this.getRequest(`users/${this.discogsUserName}/collection/releases/${releaseId}`);

        for (const collectionRelease of getReleaseInfoFromCollection.data.releases) {
            return this.postRequest(`users/${this.discogsUserName}/collection/folders/${collectionRelease.folder_id}/releases/${releaseId}/instances/${collectionRelease.instance_id}`, { folder_id: folderId });
        }
    }

    public async removeReleaseFromFolder(releaseId: string) {
        // Resets the folder to 1 - Uncategorised

        await this.calculateRateLimitRemaining();

        const getReleaseInfoFromCollection = await this.getRequest(`users/${this.discogsUserName}/collection/releases/${releaseId}`);
        console.log(getReleaseInfoFromCollection.data.releases);

        for (const collectionRelease of getReleaseInfoFromCollection.data.releases) {
            return this.postRequest(`users/${this.discogsUserName}/collection/folders/${collectionRelease.folder_id}/releases/${releaseId}/instances/${collectionRelease.instance_id}`, { folder_id: "1" });
        }
    }

//
// COLLECTION ENDPOINTS
//

    public async getUserCollection(pageNumber:string, sort:string, sortOrder:string) {
        await this.calculateRateLimitRemaining();

        if (!pageNumber){
            pageNumber="1"
        }

        if (!sortOrder){
            sortOrder="desc"
        }

        if (!sort) {
            sort = "added"
        }
        else if (sort == "label") {}
        else if (sort == "artist") {}
        else if (sort == "title") {}
        else if (sort == "catno") {}
        else if (sort == "format") {}
        else if (sort == "rating") {}
        else if (sort == "year") {}
        else{
            sort = "added"
        }

        return this.getRequest(`users/${this.discogsUserName}/collection?sort=${sort}&sort_order=${sortOrder}&per_page=${this.perPage}&page=${pageNumber}`);
    }

    public async getUserCollectionValue() {
        await this.calculateRateLimitRemaining();

        return this.getRequest(`users/${this.discogsUserName}/collection/value`);
    }

    public async listCustomFields() {
        await this.calculateRateLimitRemaining();

        return this.getRequest(`users/${this.discogsUserName}/collection/fields`);
    }

    public async editCustomFieldValue(releaseId:string, fieldId:string, value:string) {
        await this.calculateRateLimitRemaining();

        const getReleaseInfoFromCollection = await this.getRequest(`users/${this.discogsUserName}/collection/releases/${releaseId}`);

        for (const collectionRelease of getReleaseInfoFromCollection.data.releases) {
            return this.postRequest(`users/${this.discogsUserName}/collection/folders/${collectionRelease.folder_id}/releases/${releaseId}/instances/${collectionRelease.instance_id}/fields/${fieldId}`, {value: value} );
        }
    }

    public async addReleaseToCollection(releaseId: string) {
        await this.calculateRateLimitRemaining();
        const checkCollectionForRelease = await this.getRequest(`users/${this.discogsUserName}/collection/releases/${releaseId}`);
        console.log(checkCollectionForRelease);
        if (checkCollectionForRelease.data.releases.id) {
            console.log("Release In Collection")
            return {
                data: checkCollectionForRelease.data,
                headers: checkCollectionForRelease.headers,
            }
        } else {
            return this.postRequest(`users/${this.discogsUserName}/collection/folders/1/releases/${releaseId}`, {});
        }
    }

//
// WANTLIST ENDPOINTS
//

    public async getUserWantlist(pageNumber: string, sort: string, sortOrder: string) {
        await this.calculateRateLimitRemaining();

        if (!pageNumber) {
            pageNumber = "1"
        }

        if (!sortOrder) {
            sortOrder = "desc"
        }

        if (!sort) {
            sort = "added"
        }
        else if (sort == "year") { }
        else if (sort == "artist") { }
        else if (sort == "title") { }
        else if (sort == "catno") { }
        else if (sort == "format") { }
        else if (sort == "rating") { }
        else {
            sort = "added"
        }

        return this.getRequest(`users/${this.discogsUserName}/wants?sort=${sort}&sort_order=${sortOrder}&per_page=${this.perPage}&page=${pageNumber}`);
    }

    public async addToUserWantlist(releaseNumber: string) {

        return this.putRequest(`users/${this.discogsUserName}/wants/${releaseNumber}`, {})

    }

    public async removeFromUserWantlist(releaseNumber: string) {

        return this.deleteRequest(`users/${this.discogsUserName}/wants/${releaseNumber}`)

    }

//
// RELEASE ENDPOINTS
//

    public async getRelease(releaseId: string) {
        await this.calculateRateLimitRemaining();

        return this.getRequest(`releases/${releaseId}`);
    }

    public async getReleaseUserRating(releaseId: string) {
        await this.calculateRateLimitRemaining();

        return this.getRequest(`releases/${releaseId}/rating/${this.discogsUserName}`);
    }

    public async setChangeReleaseUserRating(releaseId: string, rating: number){
        await this.calculateRateLimitRemaining();

        return this.putRequest(`releases/${releaseId}/rating/${this.discogsUserName}`,{rating:rating});
    }

    public async deleteReleaseUserRating(releaseId: string){
        await this.calculateRateLimitRemaining();

        return this.deleteRequest(`releases/${releaseId}/rating/${this.discogsUserName}`);
    }

    public async getReleaseCommunityRating(releaseId: string) {
        await this.calculateRateLimitRemaining();

        return this.getRequest(`releases/${releaseId}/rating`);
    }

    public async getReleaseStats(releaseId: string) {
        await this.calculateRateLimitRemaining();

        return this.getRequest(`releases/${releaseId}/stats`);
    }

    public async getMasterRelease(masterId: string) {
        await this.calculateRateLimitRemaining();

        return this.getRequest(`masters/${masterId}`);
    }

    public async getMasterReleaseVersions(masterId: string) {
        await this.calculateRateLimitRemaining();

        return this.getRequest(`masters/${masterId}/versions`); // takes parameters, needs adding
    }

//
// ARTIST ENDPOINTS
//

    public async getArtistDetails(ArtistId: string) {
        await this.calculateRateLimitRemaining();

        return this.getRequest(`artists/${ArtistId}`);
    }

    public async getArtistReleases(ArtistId: string, pageNumber:string, sort:string, sortOrder:string) {
        await this.calculateRateLimitRemaining();

        if (!pageNumber){
            pageNumber="1"
        }

        if (!sortOrder){
            sortOrder="desc"
        }

        if (!sort) {
            sort = "title"
        }
        else if (sort == "year") {}
        else if (sort == "format") {}
        else{
            sort = "title"
        }

        return this.getRequest(`artists/${ArtistId}/releases?sort=${sort}&sort_order=${sortOrder}&per_page=${this.perPage}&page=${pageNumber}`); // takes parameters, needs adding
    }


//
// LABEL ENDPOINTS
//

    public async getLabelDetails(LabelId: string) {
        await this.calculateRateLimitRemaining();

        return this.getRequest(`labels/${LabelId}`);
    }

    public async getLabelReleases(LabelId: string, pageNumber:string, sort:string, sortOrder:string) {
        await this.calculateRateLimitRemaining();

        if (!pageNumber){
            pageNumber="1"
        }

        if (!sortOrder){
            sortOrder="desc"
        }

        if (!sort) {
            sort = "title"
        }
        else if (sort == "year") {}
        else if (sort == "artist") {}
        else if (sort == "catno"){}
        else if (sort == "format") {}
        else{
            sort = "added"
        }

        return this.getRequest(`labels/${LabelId}/releases?sort=${sort}&sort_order=${sortOrder}&per_page=${this.perPage}&page=${pageNumber}`); // takes parameters, needs adding
    }

//
// SEARCH ENDPOINTS
// 

    public getSearch(query: string, pageNumber: string) {
        if (!pageNumber) {
            pageNumber = "1"
        }
        return this.request(`database/search?q=${query}&page=${pageNumber}`);
    }

    public getSearchArtist(query: string, pageNumber: string) {
        if (!pageNumber) {
            pageNumber = "1"
        }
        return this.request(`database/search?q=${query}&type=artist&page=${pageNumber}`);
    }

    public getSearchReleaseId(query: string, pageNumber: string) {
        if (!pageNumber) {
            pageNumber = "1"
        }
        return this.request(`database/search?release_id=${query}&type=release&page=${pageNumber}`);
    }

    public getSearchCatalogueNumber(query: string, pageNumber: string) {
        if (!pageNumber) {
            pageNumber = "1"
        }
        return this.request(`database/search?catno=${query}&type=release&page=${pageNumber}`);
    }

    public getSearchReleaseTitle(query: string, pageNumber: string) {
        if (!pageNumber) {
            pageNumber = "1"
        }
        return this.request(`database/search?release_title=${query}&type=release&page=${pageNumber}`);
    }

    public getSearchTrackTitle(query: string, pageNumber: string) {
        if (!pageNumber) {
            pageNumber = "1"
        }
        return this.request(`database/search?track=${query}&type=release&page=${pageNumber}`);
    }

    public getSearchLabelReleases(query: string, pageNumber: string) {
        if (!pageNumber) {
            pageNumber = "1"
        }
        return this.request(`database/search?label=${query}&type=release&page=${pageNumber}`);
    }

    public getSearchLabel(query: string, pageNumber: string) {
        if (!pageNumber) {
            pageNumber = "1"
        }
        return this.request(`database/search?q=${query}&type=label&page=${pageNumber}`);
    }

}